ignite/pkg/cache/cache.go

```
@@ -0,0 +1,151 @@
+	dir, err := xfilepath.Join(cacheFolder, xfilepath.Path(chainName))()
+	if err != nil {
+		return Storage{}, err
+	}
+
+	return NewStorage(dir)
+}
+
+// NewStorage create a local db file (if it doesn't exist) and opens a connection to it (or reuses an existing one)
+// Usually NewChainStorage is more appropriate to use since it makes it possible to clear the cache for a particular project
+// It is safe to call multiple times
+func NewStorage(dir string) (Storage, error) {
+	db, ok := dbInstances[dir]
+	if ok {
+		return Storage{db}, nil
+	}
+
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return Storage{}, err
+	}
+
+	storagePath := filepath.Join(dir, dbName)
+	db, err := bolt.Open(storagePath, 0640, nil)
+	if err != nil {
+		return Storage{}, err
+	}
+
+	dbInstances[dir] = db
+	return Storage{db}, nil
+}
+
+// New creates a namespaced and typesafe key-value Cache
+func New[T any](storage Storage, namespace string) Cache[T] {
+	return Cache[T]{
+		storage:   storage,
+		namespace: namespace,
+	}
+}
+
+// Clear deletes all namespaces and cached values
+func (s Storage) Clear() error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		return tx.ForEach(func(name []byte, _ *bolt.Bucket) error {
+			return tx.DeleteBucket(name)
+		})
+	})
+}
+
+// Put sets key to value within the namespace
+// If the key already exists, it will be overwritten
+func (c Cache[T]) Put(key string, value T) error {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+	if err := encoder.Encode(value); err != nil {
+		return err
+	}
+	result := buf.Bytes()
+
+	return c.storage.db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(c.namespace))
+		if err != nil {
+			return err
+		}
+		return b.Put([]byte(key), result)
+	})
+}
+
+// Get fetches the value of key within the namespace.
+// If no value exists, it will return found == false
+func (c Cache[T]) Get(key string) (val T, found bool, err error) {
```

As you can see from the signature this returns 3 values, which I was hoping to avoid. I tried many different designs for the cache store, but I found this to be the cleanest in the end. I am very open to suggestions here.

Another thing I considered was if we want to actually swallow errors here or not (and just return found=false instead). It would simplify the UX of the cache module significantly, but potentially hide errors too.

---

Yeah, I did consider that as well, but I personally found the resulting code not much cleaner from this, but not a lot worse either I suppose. Let me know if you want to change this or not.

---

I like that. Making that change now!